src/sentry/incidents/subscription_processor.py

```diff
@@ -0,0 +1,198 @@
+    """
+    Class for processing subscription updates for an alert rule. Accepts a subscription
+    and then can process one or more updates via `process_update`. Keeps track of how
+    close an alert rule is to alerting, creates an incident, and auto resolves the
+    incident if a resolve threshold is set and the threshold is triggered.
+    """
+
+    REDIS_TTL = int(timedelta(days=7).total_seconds())
+    # Each entry is a tuple in format (<alert_operator>, <resolve_operator>)
+    THRESHOLD_TYPE_OPERATORS = {
+        AlertRuleThresholdType.ABOVE: (operator.gt, operator.lt),
+        AlertRuleThresholdType.BELOW: (operator.lt, operator.gt),
+    }
+    ALERT_RULE_BASE_STAT_KEY = "{alert_rule:%s}:%%s"
+    ALERT_RULE_STAT_KEYS = ("last_update", "alert_triggered", "resolve_triggered")
+
+    def __init__(self, subscription):
+        self.subscription = subscription
+        try:
+            self.alert_rule = AlertRule.objects.get(query_subscription=subscription)
+        except AlertRule.DoesNotExist:
+            return
+
+        # Fetch the active incident if one exists for this alert rule.
+        try:
+            self.active_incident = Incident.objects.filter(
+                type=IncidentType.ALERT_TRIGGERED.value,
+                status=IncidentStatus.OPEN.value,
+                alert_rule=self.alert_rule,
+            ).order_by("-date_added")[0]
+        except IndexError:
+            self.active_incident = None
+
+        self.client = self.get_redis_client()
+        self.pipeline = self.client.pipeline()
+        self.last_update, self.alert_triggers, self.resolve_triggers = self.get_alert_rule_stats()
+        self.orig_alert_triggers = self.alert_triggers
+        self.orig_resolve_triggers = self.resolve_triggers
+
+    def process_update(self, subscription_update):
+        if not hasattr(self, "alert_rule"):
+            # If the alert rule has been removed then just skip
+            metrics.incr("incidents.alert_rules.no_alert_rule_for_subscription")
+            logger.error(
+                "Received an update for a subscription, but no associated alert rule exists"
+            )
+            # TODO: Delete subscription here.
+            return
+
+        if subscription_update["timestamp"] <= self.last_update:
+            metrics.incr("incidents.alert_rules.skipping_already_processed_update")
+            return
+
+        self.last_update = subscription_update["timestamp"]
+
+        # TODO: At the moment we only have individual aggregations. Handle multiple
+        # later
+        aggregation = AlertRuleAggregations(self.alert_rule.aggregations[0])
+        aggregation_name = alert_aggregation_to_snuba[aggregation][2]
+        aggregation_value = subscription_update["values"][aggregation_name]
+
+        alert_operator, resolve_operator = self.THRESHOLD_TYPE_OPERATORS[
+            AlertRuleThresholdType(self.alert_rule.threshold_type)
+        ]
+
+        with transaction.atomic():
```

Comment:
With some of the updates being done in postgres and some being done in redis. Do you need to worry about partial/incomplete updates when transactions fail?

Reply:
I covered this a little in the pr description. Essentially the database operations will roll back if there's any issue trying to update redis. I think the main potential problem could be if we exit kill the process while updating Redis. It's possible in that case that we'll store the values in Redis, but roll back the database transaction, which would mean that we'd miss an update. I'm not sure there's a great way to work around this.